_posts/2018-12-05-graphql-union-vs-interface

```diff
@@ -0,0 +1,231 @@
+
+For example a music `Instrument` can have strings or not.  If it has strings we want to mention how many strings it has in `numberOfStrings` field. For any non-string instrument this field would be `null` in the database.
+
+One way to do this is to have `Instrument` type always have `numberOfStrings` and in the case of non-string instruments return `nil`. Sample result for this would be
+
+```json
+{
+  "data": [
+    {
+      "id": 1,
+      "name": "Guitar",
+      "numberOfStrings": 6
+    },
+    {
+      "id": 2,
+      "name": "Drums",
+      "numberOfStrings": null
+    }
+  ]
+}
+```
+
+The above solution would work, but it will add extra work on the clients to decide if `numberOfStrings` is even applicable to this current instrument or not.
+
+The more GraphLQ approach for this would be to use `Interface`. We can define a generic `Instrument`interface and have all the common fields between all instruments defined there. Then we can have each specific category of instruments define its own special fields and then access those specific fields using fragments.
+
+In `graphql-ruby`  you can define an Interface with
+
+```ruby
+module Types::InstrumentInterface
+  include Types::BaseInterface
+
+  description 'A Music Album'
+  graphql_name 'Album'
+
+  field :id, ID, null: false
+  field :name, String, null: false
+	field :category, String, null: false
+
+  definition_methods do
+    def resolve_type(object, _context)
+      case object.category
+      when "string" then Types::StringInstrument
+      when "drums" then Types::DrumInstrument
+      else
+        raise 'Unknown instrument type'
+      end
+    end
+  end
+end
+```
+
+Then we can have our specific types implementing this interface.
+
+```ruby
+class Types::StringInstrument < Types::BaseObject
+  implements Types:: InstrumentInterface
+
+  field :number_of_strings, Integer, null: false
+end
+```
+
+For types that don’t have any extra field, they can just reuse everything from interface.
+
+```ruby
+class Types::DrumInstrument < Types::BaseObject
+  implements Types:: InstrumentInterface
+end
+```
+
+This way the query for getting instruments can look like
+```ruby
+query {
+  instruments() {
+    id
+    name
+    category
+    ... on StringInstrument {
+       numberOfStrings
+    }
+  }
+}
+```
+
+TODO: sample response
+
+One issue we found after doing above was, since this way we don’t reference `StringInstrument` and `DrumInstrument` types anywhere in our schema, they actually don’t end up showing in the generated schema. For them to show up we have to add them as `orphan_types` in the interface. So the interface definition will look like
+
+```ruby
+module Types::InstrumentInterface
+  include Types::BaseInterface
+
+  description 'A Music Album'
+  graphql_name 'Album'
+
+  field :id, ID, null: false
+  field :name, String, null: false
+	field :category, String, null: false
+
+	## Changes
+  orphan_types Types::StringInstrument, Types::DrumInstrument
```

Comment:
You can use `diff` as the language specifier for backtick code blocks, and then it'll get highlighted appropriately 👍